File: exts/jphp-gui-ext/src/main/java/org/develnext/jphp/ext/javafx/classes/layout/UXVBox.java
package org.develnext.jphp.ext.javafx.classes.layout;

import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.layout.Priority;
import javafx.scene.layout.VBox;
import org.develnext.jphp.ext.javafx.JavaFXExtension;
import php.runtime.annotation.Reflection;
import php.runtime.annotation.Reflection.Property;
import php.runtime.annotation.Reflection.Signature;
import php.runtime.env.Environment;
import php.runtime.reflection.ClassEntity;

import java.util.List;

@Reflection.Name(JavaFXExtension.NS + "layout\\UXVBox")
public class UXVBox<T extends VBox> extends UXPane<VBox> {
    interface WrappedInterface {
        @Property double spacing();
        @Property Pos alignment();
        @Property boolean fillWidth();

        void requestLayout();
    }

    public UXVBox(Environment env, T wrappedObject) {
        super(env, wrappedObject);
    }

    public UXVBox(Environment env, ClassEntity clazz) {
        super(env, clazz);
    }

    @Signature
    public void __construct() {
        __wrappedObject = new VBox();
    }

    @Signature
    public void __construct(List<Node> children) {
        __construct(children, 0);
    }

    @Signature
    public void __construct(List<Node> children, double spacing) {
        __wrappedObject = new VBox(children.toArray(new Node[children.size()]));
        getWrappedObject().setSpacing(spacing);
    }

    @Signature
    public static void setVgrow(Node node, @Reflection.Nullable Priority value) {
        VBox.setVgrow(node, value);
    }

    @Signature
    public static Priority getVgrow(Node node) {
        return VBox.getVgrow(node);
    }

    @Signature
    public static void setMargin(Node node, Insets insets) {
        VBox.setMargin(node, insets);
    }
}

File: exts/jphp-gui-ext/src/main/resources/JPHP-INF/sdk/php/gui/layout/UXVBox.php
<?php
namespace php\gui\layout;

use php\gui\UXNode;

/**
 * Class UXVBox
 * @package php\gui\layout
 * @packages gui, javafx
 */
class UXVBox extends UXPane
{
    /**
     * TOP_LEFT, TOP_CENTER, TOP_RIGHT, CENTER_LEFT, ... CENTER, ... BOTTOM_RIGHT,
     * BASELINE_LEFT, BASELINE_CENTER, BASELINE_RIGHT
     * @var string
     */
    public $alignment;

    /**
     * @var float
     */
    public $spacing;

    /**
     * @var bool
     */
    public $fillWidth = true;

    /**
     * @param UXNode[] $nodes [optional]
     * @param float|int $spacing
     */
    public function __construct(array $nodes, $spacing = 0)
    {
    }

    public function requestLayout()
    {
    }

    /**
     * @param UXNode $node
     * @param string $value
     */
    static public function setVgrow(UXNode $node, $value)
    {
    }

    /**
     * @param UXNode $node
     * @return string
     */
    static public function getVgrow(UXNode $node)
    {
    }

    /**
     * @param UXNode $node
     * @param double[]|double $margins
     */
    static public function setMargin(UXNode $node, $margins)
    {
    }
}